### PR TITLE
Convert formatting features to AnalyzerConfigOptions

### PR DESCRIPTION
**Pull request by sharwell**
I predicted three points of most likely failure, which have been addressed as described:

* Options that do not have **.editorconfig** storage will always have the default value in `AnalyzerConfigOptions`. An assertion is now applied to catch these in debug builds.
* Options that are not registered through `IOptionProvider` will always have the default value in the `AnalyzerConfigOptions` instance provided by `OptionSet.AsAnalyzerConfigOptions`. An assertion is now applied to catch these in debug builds.
* Options that do not round-trip with full fidelity may have loss of precision. I have not done anything to mitigate this situation, and it seems to be much less likely to occur.

### CODE REVIEW

File: src/CodeStyle/Core/Analyzers/Options/OptionSetExtensions.cs [hunk 1]
```diff
@@ -19,16 +19,22 @@ public static T GetOption<T>(this OptionSet optionSet, Option<T> option)
...
         {
             if (!TryGetEditorConfigOption(optionSet, option, out T value))
-            {
                 value = option.DefaultValue;
-            }
 
             return value;
         }
 
+        public static T GetOption<T>(this OptionSet optionSet, PerLanguageOption<T> option, string language)
```
Comment (mavasani) [inline src/CodeStyle/Core/Analyzers/Options/OptionSetExtensions.cs]: Why do you need this method? Are there too many callsites to be adjusted to remove the language argument?

Comment (sharwell) [reply src/CodeStyle/Core/Analyzers/Options/OptionSetExtensions.cs]: It's needed for cases where we still conditionally use `AnalyzerConfigOptions` for Code Style and `OptionSet` for Workspaces.

Comment (mavasani) [reply src/CodeStyle/Core/Analyzers/Options/OptionSetExtensions.cs]: I am assuming we want to eventually remove all such cases, at which point this can go away?

Comment (sharwell) [reply src/CodeStyle/Core/Analyzers/Options/OptionSetExtensions.cs]: Yes that's what I'm hoping.

File: src/Workspaces/Core/Portable/Options/EditorConfig/IEditorConfigStorageLocation2.cs [hunk 2]
```diff
@@ -24,5 +24,10 @@ internal interface IEditorConfigStorageLocation2 : IEditorConfigStorageLocation
         /// Gets the editorconfig string representation for this storage location.
         /// </summary>
         string GetEditorConfigString(object? value, OptionSet optionSet);
+
+        /// <summary>
+        /// Gets the editorconfig string representation for this storage location. The result does not include the key.
+        /// </summary>
```
Comment (mavasani) [inline src/Workspaces/Core/Portable/Options/EditorConfig/IEditorConfigStorageLocation2.cs]: Probably include an example in the doc comment for this API and the previous API `GetEditorConfigString` to add clarity?

Comment (sharwell) [reply src/Workspaces/Core/Portable/Options/EditorConfig/IEditorConfigStorageLocation2.cs]: Done

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/AnalyzerConfigOptionsExtensions.cs [hunk 3]
```diff
@@ -10,22 +11,34 @@
...
+            return GetOption(analyzerConfigOptions, option);
+        }
+#endif
+
+        public static T GetOption<T>(this AnalyzerConfigOptions analyzerConfigOptions, Option<T> option)
         {
-            if (!TryGetEditorConfigOption(optionSet, option, out T value))
+            if (!TryGetEditorConfigOption(analyzerConfigOptions, option, out T value))
             {
+                Debug.Fail("Failed to find a .editorconfig key for the option.");
```
Comment (mavasani) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/AnalyzerConfigOptionsExtensions.cs]: I like this - this should prevent us from accidentally querying non-editorconfig based option from CodeStyle layer.

File: src/EditorFeatures/CSharp/AutomaticCompletion/Sessions/CurlyBraceCompletionSession.cs [hunk 4]
```diff
@@ -108,7 +110,8 @@ private static bool ContainsOnlyWhitespace(IBraceCompletionSession session)
 
         private IEnumerable<AbstractFormattingRule> GetFormattingRules(Document document)
         {
-            return SpecializedCollections.SingletonEnumerable(BraceCompletionFormattingRule.Instance).Concat(Formatter.GetDefaultFormattingRules(document));
+            var indentStyle = document.GetOptionsAsync(CancellationToken.None).WaitAndGetResult_CanCallOnBackground(CancellationToken.None).GetOption(FormattingOptions.SmartIndent);
```
Comment (mavasani) [inline src/EditorFeatures/CSharp/AutomaticCompletion/Sessions/CurlyBraceCompletionSession.cs]: Do we need a synchronous `GetOptions` or `GetOption` helper here or you think that will likely be misused?

File: src/Workspaces/Core/Portable/Formatting/Formatter.cs [hunk 5]
```diff
@@ -246,10 +246,13 @@ internal static IFormattingResult GetFormattingResult(SyntaxNode node, IEnumerab
...
             }
 
+            var optionService = workspace.Services.GetRequiredService<IOptionService>();
+
             options ??= workspace.Options;
             rules ??= GetDefaultFormattingRules(workspace, node.Language);
             spans ??= SpecializedCollections.SingletonEnumerable(node.FullSpan);
-            return languageFormatter.Format(node, spans, options, rules, cancellationToken);
+            var shouldUseFormattingSpanCollapse = options.GetOption(FormattingOptions.AllowDisjointSpanMerging);
+            return languageFormatter.Format(node, spans, shouldUseFormattingSpanCollapse, options.AsAnalyzerConfigOptions(optionService, node.Language), rules, cancellationToken);
```
Comment (mavasani) [inline src/Workspaces/Core/Portable/Formatting/Formatter.cs]: Why this change of explicitly querying `FormattingOptions.AllowDisjointSpanMerging` and pass it in? I presume invoking GetOption on the passed in AnalyzerConfigOptions argument should give the same value to the callee given that the wrapping AnalyzerConfigOptions implementation falls back to the original option set?

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/DefaultOperationProvider.cs [hunk 6]
```diff
@@ -3,14 +3,9 @@
 // See the LICENSE file in the project root for more information.
 
 using System.Collections.Generic;
+using Microsoft.CodeAnalysis.Diagnostics;
 using Microsoft.CodeAnalysis.Formatting.Rules;
 
-#if CODE_STYLE
```
Comment (mavasani) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/DefaultOperationProvider.cs]: Nice!

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Rules/ElasticTriviaFormattingRule.cs [hunk 7]
```diff
@@ -12,20 +12,15 @@
 using Microsoft.CodeAnalysis.Formatting.Rules;
 using Microsoft.CodeAnalysis.Shared.Extensions;
 using Microsoft.CodeAnalysis.Shared.Utilities;
-
-#if CODE_STYLE
-using OptionSet = Microsoft.CodeAnalysis.Diagnostics.AnalyzerConfigOptions;
-#else
-using Microsoft.CodeAnalysis.Options;
-#endif
+using Microsoft.CodeAnalysis.Diagnostics;
```
Comment (mavasani) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Rules/ElasticTriviaFormattingRule.cs]: Nit: sort usings

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Rules/IndentBlockFormattingRule.cs [hunk 8]
```diff
@@ -11,30 +11,25 @@
 using Microsoft.CodeAnalysis.Shared.Extensions;
 using Microsoft.CodeAnalysis.Shared.Utilities;
 using Microsoft.CodeAnalysis.Text;
-
-#if CODE_STYLE
-using OptionSet = Microsoft.CodeAnalysis.Diagnostics.AnalyzerConfigOptions;
-#else
-using Microsoft.CodeAnalysis.Options;
-#endif
+using Microsoft.CodeAnalysis.Diagnostics;
```
Comment (mavasani) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Rules/IndentBlockFormattingRule.cs]: Nit: sort

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Formatting/TriviaEngine/AbstractTriviaFormatter.cs [hunk 9]
```diff
@@ -10,12 +10,7 @@
 using Microsoft.CodeAnalysis.Shared.Extensions;
 using Microsoft.CodeAnalysis.Text;
 using Roslyn.Utilities;
-
-#if CODE_STYLE
-using OptionSet = Microsoft.CodeAnalysis.Diagnostics.AnalyzerConfigOptions;
-#else
-using Microsoft.CodeAnalysis.Options;
-#endif
+using Microsoft.CodeAnalysis.Diagnostics;
```
Comment (mavasani) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Formatting/TriviaEngine/AbstractTriviaFormatter.cs]: Nit: sort

File: src/EditorFeatures/CSharp/AutomaticCompletion/Sessions/CurlyBraceCompletionSession.cs (hunk 4 above)
Comment (sharwell) [reply src/EditorFeatures/CSharp/AutomaticCompletion/Sessions/CurlyBraceCompletionSession.cs]: I'm hoping to see `document.GetOptionsAsync` reduce in usage over time so most of these cases will naturally go away.

File: src/Workspaces/Core/Portable/Formatting/Formatter.cs (hunk 5 above)
Comment (sharwell) [reply src/Workspaces/Core/Portable/Formatting/Formatter.cs]: `AllowDisjointSpanMerging` isn't backed by **.editorconfig**, so there is no way to provide it through `AnalyzerConfigOptions`.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Formatting/TriviaEngine/AbstractTriviaFormatter.cs (hunk 9 above)
Comment (sharwell) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Formatting/TriviaEngine/AbstractTriviaFormatter.cs]: done

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Rules/IndentBlockFormattingRule.cs (hunk 8 above)
Comment (sharwell) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Rules/IndentBlockFormattingRule.cs]: done

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Rules/ElasticTriviaFormattingRule.cs (hunk 7 above)
Comment (sharwell) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Rules/ElasticTriviaFormattingRule.cs]: done
